File: tests/test_live_context.py
import copy
import json
import unittest
from pathlib import Path

from catcher_intel.live_context import game_context


class GameContextTests(unittest.TestCase):
    def setUp(self):
        self.feed = json.loads((Path(__file__).parent / "fixtures/mlb-game.json").read_text())

    def test_final_preserves_unplayed_half_inning(self):
        context = game_context(self.feed)
        self.assertEqual(context["teams"]["home"]["runs"], 12)
        self.assertEqual(context["teams"]["away"]["runs"], 2)
        self.assertIsNone(context["innings"][-1]["home"])
        self.assertEqual(context["decisions"]["winner"]["name"], "Shota Imanaga")

    def test_preview_does_not_invent_scores_or_players(self):
        self.feed["liveData"] = {}
        context = game_context(self.feed)
        self.assertIsNone(context["teams"]["home"]["runs"])
        self.assertIsNone(context["matchup"]["catcher"]["id"])
        self.assertEqual(context["recent_plays"], [])
        self.assertIsNotNone(context["probable_pitchers"]["home"]["name"])

    def test_current_players_and_runners_come_from_field_state(self):
        feed = copy.deepcopy(self.feed)
        line = feed["liveData"]["linescore"]
        line["offense"]["first"] = {"id": 123, "fullName": "Test Runner"}
        line["offense"]["batter"] = {"id": 124, "fullName": "Next Batter"}
        context = game_context(feed)
        self.assertEqual(context["runners"]["first"]["name"], "Test Runner")
        self.assertIsNone(context["runners"]["second"]["id"])
        self.assertEqual(context["matchup"]["batter"]["name"], "Next Batter")
        self.assertEqual(context["matchup"]["catcher"]["name"], "Carson Kelly")


if __name__ == "__main__":
    unittest.main()

File: apps/web/scripts/test-live-context.mjs
import assert from "node:assert/strict";
import fs from "node:fs";
import { execFileSync } from "node:child_process";
import ts from "typescript";
import { fileURLToPath } from "node:url";

const root = new URL("../../../", import.meta.url);
const source = fs.readFileSync(new URL("../src/lib/live-game-context.ts", import.meta.url), "utf8");
const compiled = ts.transpileModule(source, { compilerOptions: { module: ts.ModuleKind.ESNext, target: ts.ScriptTarget.ES2022 } }).outputText;
const { buildGameContext, mlbDate, shiftDate } = await import(`data:text/javascript;base64,${Buffer.from(compiled).toString("base64")}`);
const fixture = JSON.parse(fs.readFileSync(new URL("tests/fixtures/mlb-game.json", root), "utf8"));

assert.equal(mlbDate(new Date("2026-09-12T02:00:00Z")), "2026-09-11");
assert.equal(shiftDate("2026-03-09", -1), "2026-03-08");
assert.equal(shiftDate("2026-01-01", -1), "2025-12-31");
const context = buildGameContext(fixture);
assert.equal(context.teams.home.runs, 12);
assert.equal(context.innings.at(-1).home, null);
assert.equal(context.decisions.winner.name, "Shota Imanaga");
assert.equal(buildGameContext({}).teams.home.runs, null);

// The same source feed must render identically with and without FastAPI.
for (const feed of [fixture, {}, { ...fixture, liveData: {} }, {
  ...fixture, liveData: { ...fixture.liveData, linescore: { ...fixture.liveData.linescore,
    offense: { first: { id: 123, fullName: "Test Runner" }, batter: { id: 124, fullName: "Next Batter" } },
  } },
}]) {
  const python = execFileSync(fileURLToPath(new URL(".venv/bin/python", root)), ["-c", "import sys,json; from catcher_intel.live_context import game_context; print(json.dumps(game_context(json.load(sys.stdin))))"], { input: JSON.stringify(feed), encoding: "utf8" });
  assert.deepEqual(buildGameContext(feed), JSON.parse(python));
}
console.log("PASS MLB feed mapping, missing data, Eastern dates, and Python/frontend parity");
